lib/events.js

```diff
@@ -185,89 +365,101 @@ EventEmitter.prototype.emit = function emit(type) {
+          events = this._events;
+        }
+        existing = events[type];
+      }
 
-  if (!existing) {
-    // Optimize the case of one listener. Don't need the extra array object.
-    existing = events[type] = listener;
-    ++this._eventsCount;
-  } else {
-    if (typeof existing === 'function') {
-      // Adding the second element, need to change to array.
-      existing = events[type] = [existing, listener];
-    } else {
-      // If we've already got an array, just append.
-      existing.push(listener);
-    }
+      if (!existing) {
+        // Optimize the case of one listener. Don't need the extra array object.
+        if (once)
+          existing = events[type] = { once: listener, fired: false };
+        else
+          existing = events[type] = listener;
+        ++this._eventsCount;
+      } else {
+        var existingOnce = existing.once;
+        if (typeof existing === 'function' || existingOnce) {
+          // Adding the second element, need to change to array.
+          if (once) {
+            existing = events[type] = [
+              existing,
+              { once: listener, fired: false }
+            ];
+            existing.onceCount = (existingOnce ? 2 : 1);
+          } else {
+            existing = events[type] = [existing, listener];
+            existing.onceCount = (existingOnce ? 1 : 0);
+          }
+        } else {
+          // If we've already got an array, just append.
+          if (once) {
+            existing.push({ once: listener, fired: false });
+            ++existing.onceCount;
+          } else
+            existing.push(listener);
+        }
 
-    // Check for listener leak
-    if (!existing.warned) {
-      m = $getMaxListeners(this);
-      if (m && m > 0 && existing.length > m) {
-        existing.warned = true;
-        console.error('(node) warning: possible EventEmitter memory ' +
-                      'leak detected. %d %s listeners added. ' +
-                      'Use emitter.setMaxListeners() to increase limit.',
-                      existing.length, type);
-        console.trace();
+        // Check for listener leak
+        if (!existing.warned) {
+          m = $getMaxListeners(this);
+          if (m && m > 0 && existing.length > m) {
+            existing.warned = true;
+            console.error('(node) warning: possible EventEmitter memory ' +
+                          'leak detected. %d %s listeners added. ' +
+                          'Use emitter.setMaxListeners() to increase limit.',
+                          existing.length, type);
+            console.trace();
+          }
+        }
       }
-    }
-  }
 
-  return this;
+      return this;
+    };
+
+EventEmitter.prototype.on = function on(type, listener) {
+  if (!this._addListener) {
+    return EventEmitter.prototype._addListener.call(this, type, listener,
+                                                    false);
+  }
+  return this._addListener(type, listener, false);
```

Comment:
Since people inherit from the `EventEmitter` the object map for the `this` will change. Which means the arguments map will also change in the call to `_addListener()`. It's better if the call was made on the prototype of the object.